pinot-core/src/main/java/com/linkedin/pinot/core/periodictask/PeriodicTaskScheduler.java

```diff
@@ -15,58 +15,63 @@
 
   /**
    * Start scheduling periodic tasks.
    */
   public void start(List<PeriodicTask> periodicTasks) {
-    if (periodicTasks == null || periodicTasks.isEmpty()) {
-      LOGGER.warn("No periodic task assigned to scheduler!");
-      return;
+    if (_executorService != null) {
+      LOGGER.warn("Periodic task scheduler already started");
     }
 
-    if (periodicTasks.size() > CORE_POOL_SIZE) {
-      LOGGER.warn("The number of tasks:{} is more than the default number of threads:{}.", periodicTasks.size(),
-          CORE_POOL_SIZE);
+    List<PeriodicTask> tasksWithValidInterval = new ArrayList<>();
+    for (PeriodicTask periodicTask : periodicTasks) {
+      if (periodicTask.getIntervalInSeconds() > 0) {
+        tasksWithValidInterval.add(periodicTask);
+      }
     }
 
-    LOGGER.info("Starting PeriodicTaskScheduler.");
-    // Set up an executor that executes tasks periodically
-    for (PeriodicTask periodicTask : periodicTasks) {
-      periodicTask.init();
-      _executorService.scheduleWithFixedDelay(() -> {
-        try {
-          periodicTask.run();
-        } catch (Throwable e) {
-          // catch all errors to prevent subsequent executions from being silently suppressed
-          // Ref: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledExecutorService.html#scheduleWithFixedDelay-java.lang.Runnable-long-long-java.util.concurrent.TimeUnit-
-          LOGGER.warn("Caught exception while running Task: {}", periodicTask.getTaskName(), e);
-        }
-      }, periodicTask.getInitialDelayInSeconds(), periodicTask.getIntervalInSeconds(), TimeUnit.SECONDS);
+    if (tasksWithValidInterval.isEmpty()) {
+      LOGGER.warn("No periodic task scheduled");
+    } else {
+      LOGGER.info("Starting periodic task scheduler with tasks: {}", tasksWithValidInterval);
+      _executorService = Executors.newScheduledThreadPool(tasksWithValidInterval.size());
+      for (PeriodicTask periodicTask : periodicTasks) {
+        periodicTask.init();
+        _executorService.scheduleWithFixedDelay(() -> {
+          try {
+            periodicTask.run();
+          } catch (Throwable e) {
+            // catch all errors to prevent subsequent executions from being silently suppressed
+            // Ref: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledExecutorService.html#scheduleWithFixedDelay-java.lang.Runnable-long-long-java.util.concurrent.TimeUnit-
+            LOGGER.warn("Caught exception while running Task: {}", periodicTask.getTaskName(), e);
+          }
+        }, periodicTask.getInitialDelayInSeconds(), periodicTask.getIntervalInSeconds(), TimeUnit.SECONDS);
+      }
     }
   }
 
   public void stop() {
-    LOGGER.info("Stopping PeriodicTaskScheduler");
-    _executorService.shutdown();
+    if (_executorService != null) {
+      LOGGER.info("Stopping periodic task scheduler");
+      _executorService.shutdown();
+      _executorService = null;
```

Comment:
Do you need to await termination of tasks?

Reply:
That could be very slow. Once we shut down the executor, no new task will be scheduled.